### DISCUSSION TITLE
How to support both webP and JPG?

### DISCUSSION THREAD

**Discussion by jacktan165** [strapi/strapi#7559] [Legacy Topics]
When users upload a JPG image, it gets uploaded onto AWS S3. I want that image to be in webP for performance reasons. Is there a way for Strapi to support multiple extensions?

**Comment by derrickmehaffy** [comment]
Can you clarify the question? Are you wanting Strapi to convert the JPG image to WebP or is this handled externally and you just want to update the reference to the file within the database?

**Comment by amirc** [comment]
I don't know what was the author attention. But I'm interested in the first.
That is, I would like Strapi to convert the uploaded images to WebP format and will provide it as an additional parameter (like the [small, medium, large](https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/services/image-manipulation.js#L85)).
I may have time to extend the behavior, but I would love to get some guidance.

**Reply by lauriejim** [reply]
Hello! If you want to customize this function, you can use the [customization concept](https://strapi.io/documentation/v3.x/concepts/customization.html#plugin-extensions).
And you will have to update the `optimize` function. You receive a buffer then manage the thing as you want.

**Reply by robots4life** [reply]
Would you be able to elaborate on this with a simple example of the image-manipulation.js with all breakpoints also being encoded to WebP? Pretty new to Strapi and Node, managed to get a few things going but not sure how to run a Sharp instance over all the breakpoints. Also, how can I pass encoding options, i.e. quality: 80 or lossless: ture? If anyone has a bit of time to go into this that would be great, thank you heaps.

**Reply by derrickmehaffy** [reply]
Hello,

I see you are wanting to ask a question, we are currently planning to decomission our GitHub Discussions in early 2021 and we don't want to lose the great questions asked by our community.

Questions should be directed to [our forum](https://forum.strapi.io/). Please see the following contributing guidelines for asking a question [here](https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md#reporting-an-issue).

**Comment by amirc** [comment]
Cool thanks!

**Reply by robots4life** [reply]
Did you manage to get this going? If so would you be able to kindly let me know how please? Can you post your example of the image-manipulation.js please? Like I said, very new to Strapi and pretty new to Node, managed to get a few things going but not sure how to run a Sharp instance over the breakpoints. Any help or an example would really help me. Thank you.

**Comment by tommypratama** [comment]
I recommend using cloudinary to do this. Because they automatically perform image optimization and automatically convert to `webP` if the browser allows.

However, if you still want to use AWS S3. You can use these plugins:

- https://www.npmjs.com/package/strapi-provider-upload-aws-s3-with-webp

**Reply by robots4life** [reply]
@tommypratama Thank you. Yeah cloudinary is an option, though not every client wants to use this, but it surly is an option. :+1: 

I had a look at https://strapi.io/documentation/v3.x/plugins/upload.html#models-definition and https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#constructor-property as well as https://github.com/gfritzdev/strapi-provider-upload-aws-s3-with-webp/blob/master/lib/index.js

Am I correct assuming that if I want to have custom breakpoints and a custom file naming pattern I would have to do this inside https://github.com/gfritzdev/strapi-provider-upload-aws-s3-with-webp/blob/master/lib/index.js **OR** do I need to do something like this https://github.com/sarpisik/strapi-customize-image-formats/issues/1 where I have control over what sizes and file names are being created? :thinking: In fact I am wondering where and at what stage of the upload I need to inject the config I want.

@derrickmehaffy Kindly excuse my reply here, once I have managed to establish a workflow where I have control over breakpoints/sizes and file names and how to up that on AWS S3 I shall write a detailed post in the forum for sure. I am missing this information and images being such an important part of any app or site I will try and help others searching for this info. If there is a tut about this or if this is being covered in the Strapi Academy then do kindly link me to this, trying to collect all info on this while also learning this. Thank you. :pray: :heart:
